lua/tardis/libraries/sh_settings.lua

```diff
@@ -80,66 +49,236 @@ function TARDIS:GetSettingsData()
+    elseif IsEntity(src) then
+        ply = src:GetCreator()
+    end
+
+    if not id then error("Requested setting with no id") end
+    local data = self.SettingsData[id]
+    if not data then error("Requested setting " .. id .. " does not exist") end
+
+    local function select_return_val(table_value)
+        if table_value ~= nil then
+            return table_value
+        end
+        return data.value
+    end
+
+    if data.class == "global" then
+        return select_return_val(self.GlobalSettings[id])
+    end
+
+    if data.class == "local" then
+        if SERVER then
+            error("Local setting " .. id .. "is being requested serverside")
+        end
+        return select_return_val(self.LocalSettings[id])
+    end
+
+    if data.class == "networked" then
+
+        if CLIENT and (ply == nil or ply == LocalPlayer()) then
+            return select_return_val(self.NetworkedSettings[id])
+        end
+
+        if IsValid(ply) then
+            if self.ClientSettings[ply:UserID()] then
+                return select_return_val(self.ClientSettings[ply:UserID()][id])
+            end
+            return select_return_val(nil)
+        end
+
+        error("Networked setting " .. id .. " was requested for invalid player " .. tostring(ply))
     end
+
+    error("Requested setting " .. "either doesn't exist or has no defined class")
+    return
 end
 
-function TARDIS:SendSetting(id,value,ply)
-    net.Start(SERVER and "TARDIS-Settings" or "TARDIS-ClientSettings")
-        net.WriteBool(true)
-        net.WriteType(id)
-        net.WriteType(value)
+--------------------------------------------------------------------------------
+-- Saving
+
+local LOCAL_SETTINGS_FILE = "tardis_settings_cl.txt"
+local NETWORKED_SETTINGS_FILE = "tardis_settings_cl_nw.txt"
+local GLOBAL_SETTINGS_FILE = "tardis_settings_sv.txt"
+
+function TARDIS:SaveSettings()
+
+    local function SaveSettingsToFile(settings_table, settings_file)
+        local settings={}
+        for k,v in pairs(settings_table) do
+            local data = self.SettingsData[k]
+            if data and data.value ~= v then
+                settings[k] = v
+            end
+        end
+        file.Write(settings_file, self.von.serialize(settings))
+    end
+
     if SERVER then
-        if IsValid(ply) then
-            net.Send(ply)
-        else
-            net.Broadcast()
+        SaveSettingsToFile(self.GlobalSettings, GLOBAL_SETTINGS_FILE)
+    else
+        SaveSettingsToFile(self.LocalSettings, LOCAL_SETTINGS_FILE)
+        SaveSettingsToFile(self.NetworkedSettings, NETWORKED_SETTINGS_FILE)
+    end
+end
+
+function TARDIS:LoadSettings()
+
+    local function LoadSettingsFromFile(settings_table, settings_file)
+        if file.Exists(settings_file, "DATA") then
+            local file_contents = file.Read(settings_file, "DATA")
+            local loaded_settings = self.von.deserialize(file_contents)
+
+            table.Merge(settings_table, loaded_settings)
```

Comment:
How does this work if we are converting the settings as key value above when merging back in for load?

Reply:
Honestly, I have no idea
I haven't changed the way they are saved or loaded - just refactored the process

Reply:
Well it seems to work so whatever I guess lol